### PR TITLE
Issue #137: Refactored Retry so that it can be shared and used for mu…

### PR DESCRIPTION
**Pull request by RobWin**
…ltiple requests.

### CODE REVIEW

File: resilience4j-metrics/src/test/java/io/github/resilience4j/metrics/TimerTest.java [hunk 1]
```diff
@@ -104,22 +104,22 @@ public void shouldExecuteCallable() throws Throwable {
...
-        // Then the helloWorldService should be invoked 1 time
+        // Then the helloWorldService should be invoked 1 context
         BDDMockito.then(helloWorldService).should(times(1)).returnHelloWorldWithException();
     }
 
 
     @Test
     public void shouldDecorateRunnable() throws Throwable {
-        // And measure the time with Metrics
+        // And measure the context with Metrics
```
Comment (jkwatson) [inline resilience4j-metrics/src/test/java/io/github/resilience4j/metrics/TimerTest.java]: I think something may have gone wrong with a rename...

**Comment by RobWin** [comment]
I have to adapt the metrics Module when the changes are approved.

**Comment by jkwatson** [comment]
Would love to see the thread-safety documentation added to the `Retry` interface, while you're in here.

**Comment by RobWin** [comment]
Tests are green now

**Comment by RobWin** [comment]
@drmaas Can you also have a look at the ratpack Module changes. Tests are green.

**Comment by RobWin** [comment]
Ok. I'm done now

File: resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java [hunk 2]
```diff
@@ -17,9 +17,11 @@
 public interface Timer {
 
     /**
-     * Starts the Timer
+     * Creates a Timer context and starts the timer
+     *
+     * @return the Timer context
      */
-    Context time();
+    Context context();
```
Comment (storozhukBM) [inline resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java]: I see too manny things called “context”. 
It is vary hard to distinguish them if you see this code first time.
I’d suggest to think about other names.

File: resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java [hunk 3]
```diff
@@ -126,7 +128,7 @@ default void executeRunnable(Runnable runnable) throws Exception {
      */
     static <T> CheckedFunction0<T> decorateCheckedSupplier(Timer timer, CheckedFunction0<T> supplier){
         return () -> {
-            final Context context = timer.time();
+            final Context context = timer.context();
```
Comment (storozhukBM) [inline resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java]: There is possibility to make `Timer` garbage free as we did it with `CircuitBreaker`.
It requires to changes contract of `Timer.onSuccess` and `Timer.onError`
We have already discussed this in Issue #117 
This is just suggestion, you can skip it if you want to.

File: resilience4j-retry/src/main/java/io/github/resilience4j/retry/internal/AsyncRetryContext.java [hunk 4]
```diff
@@ -23,18 +25,60 @@
...
+        succeededAfterRetryCounter = new LongAdder();
+        failedAfterRetryCounter = new LongAdder();
+        succeededWithoutRetryCounter = new LongAdder();
+        failedWithoutRetryCounter = new LongAdder();
+    }
+
+    public final class ContextImpl implements AsyncRetry.Context {
+
+        private final AtomicInteger numOfAttempts = new AtomicInteger(0);
+        private AtomicReference<Throwable> lastException = new AtomicReference<>();
```
Comment (storozhukBM) [inline resilience4j-retry/src/main/java/io/github/resilience4j/retry/internal/AsyncRetryContext.java]: `lastException` should also be `final`

File: resilience4j-retry/src/main/java/io/github/resilience4j/retry/internal/RetryContext.java [hunk 5]
```diff
@@ -32,110 +33,136 @@
...
-     */
-    @Override
-    public void onSuccess() {
-        int currentNumOfAttempts = numOfAttempts.get();
-        if(currentNumOfAttempts > 0){
-            Throwable throwable = Option.of(lastException.get()).getOrElse(lastRuntimeException.get());
-            publishRetryEvent(() -> new RetryOnSuccessEvent(getName(), currentNumOfAttempts, throwable));
+        private final AtomicInteger numOfAttempts = new AtomicInteger(0);
+        private AtomicReference<Exception> lastException = new AtomicReference<>();
+        private AtomicReference<RuntimeException> lastRuntimeException = new AtomicReference<>();
```
Comment (storozhukBM) [inline resilience4j-retry/src/main/java/io/github/resilience4j/retry/internal/RetryContext.java]: lastException  and lastRuntimeException  can be `final`

File: resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java (hunk 3 above)
Comment (storozhukBM) [inline resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java]: We can also make it in scope of other issue.

Comment (RobWin) [inline resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java]: Better scope of another issue

File: resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java (hunk 2 above)
Comment (RobWin) [inline resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java]: We could rename `TimerContext` to `TimerImpl`.

Comment (RobWin) [inline resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java]: Or better I change the method name back to `time()`

Comment (storozhukBM) [inline resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java]: 👍

Comment (storozhukBM) [inline resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java]: And left `time` name also.

Comment (RobWin) [inline resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java]: left?

Comment (storozhukBM) [inline resilience4j-metrics/src/main/java/io/github/resilience4j/metrics/Timer.java]: Ohh, sorry I mean previous `time()` method name should remain.

File: resilience4j-retry/src/main/java/io/github/resilience4j/retry/Retry.java [hunk 6]
```diff
@@ -29,6 +29,10 @@
 import java.util.function.Function;
 import java.util.function.Supplier;
 
+/**
+ * A Retry instance is thread-safe can be used to decorate multiple requests.
```
Comment (jkwatson) [inline resilience4j-retry/src/main/java/io/github/resilience4j/retry/Retry.java]: 👍

**Pull request merged**